posts/2018-05-09-tagless-final-streaming.md

```diff
@@ -0,0 +1,172 @@
+---
+layout: post
+title: Tagless Final algebras and Streaming
+
+meta:
+  nav: blog
+  author: gvolpe
+  pygments: true
+
+tut:
+  scala: 2.12.4
+  binaryScala: "2.12"
+  dependencies:
+    - org.scala-lang:scala-library:2.12.4
+    - org.typelevel::cats-core:1.1.0
+
+---
+
+There have been a couple of really [nice blog posts](https://typelevel.org/blog/2017/12/27/optimizing-final-tagless.html) about `Tagless Final` and some related topics. However, I have faced some design problems when writing some algebras and haven't seen anybody talking about. So please let me introduce this problem to you.
+
+### Algebra definition
+
+Consider the following algebra:
+
+```scala
+trait ItemRepository[F[_]] {
+  def findAll: F[List[Item]]
+  def find(name: ItemName): F[Option[Item]]
+  def save(item: Item): F[Unit]
+  def remove(name: ItemName): F[Unit]
+}
+```
+
+Let's go through each method's definition:
+
+- `findAll` needs to return many Items, obtainable inside a context: `F[List[Item]]`.
+- `find` might or might not return an Item inside a context: `F[Option[Item]]`.
+- `save` and `remove` will perform some actions without returning any actual value: `F[Unit]`.
+
+Everything is clear and you might have seen this kind of pattern before, so let's create an interpreter for it:
+
+```scala
+import cats.effect.Sync
+
+// Pretend this is the real implementation... What matters for the example are the types.
+class PostgreSQLItemRepository[F[_]](implicit F: Sync[F]) extends ItemRepository[F] {
+  private val mockItem = Item(ItemName("laptop"), BigDecimal(1500))
+
+  override def findAll: F[List[Item]] = F.pure(List(mockItem))
+  override def find(name: ItemName): F[Option[Item]] = F.pure(Some(mockItem))
+  override def save(item: Item): F[Unit] = F.unit
+  override def remove(name: ItemName): F[Unit] = F.unit
+}
```

Comment:
This makes a poor example for having `Sync` constraint. Even if it's just a stub, you can add some `println`s to show that there are effects you suspend. Right now it only requires `Applicative` methods

Especially considering a latter point about abstracting over effect types.

Reply:
Fair point, I guess I was a bit lazy with this example, will update to have at least `F.delay(println(s"saving item: $item"))` :smile: